### ISSUE TITLE
get-env should not return unset!

### ISSUE THREAD

**Issue by gregit** [metaeducation/ren-c#222]
While working on a Web CGI, I'm doing:

```
val: probe get-env "QUERY_STRING"
```

In case there is no query string, this returns:

```
unset! ** Script error: val: needs a value ** Where: do either either either finish-rl-start ** Near: do intern code
```

Not sure this is expected behavior. Returning none would be better.

**Comment by hostilefork** [comment]
The GET and GET/OPT *(formerly get/any)* pattern is based on the idea that you want to be explicit about being able to handle no result, that locates the point of failure more certainly.

Might it be better to follow the pattern set by Rebol's GET and offer an /OPT so that you can say you don't care if it's there or not?  That gives two ways to deal with it--processing the unset or not.

A nice thing about returning unset is interaction with refinement revocation:

     foo: func [stuff /query query-string] [
         ; let's say this can do something meaningful without a query string
     ]

     foo/query {some stuff} (get-env/opt "QUERY_STRING")

Under refinement revocation, get-env/opt returning unset would act "as if" the /query refinement had not been provided when it called foo.

**Comment by gregit** [comment]
Yes, a get-env/opt would be great! Better than returning none.

**Comment by hostilefork** [comment]
Note that this wouldn't solve your question, because it would be "even worse" in that GET-ENV would deliver an error inside of it that "QUERY_STRING" is not available in the environment.  So you would be needing the /OPT switch to get the behavior today that lets you process the result at all.

I'm not so much "answering your question" as presenting a possibility of following the pattern of Rebol's GET and GET/OPT (a.k.a. get/any)   What is nice about it is that it lets you write your code assuming everything is there and then giving error messages if it isn't.  So here we would be treating the environment kind of like words, and making the problems come up similarly.

I like it, because I like [the CHECK-SET idea](https://trello.com/c/DM0OdyKX) and I like the idea of being able to read code and know whether something is expected to be there or not.  It's more explicit.

By the way, if you are working on your code and don't mind scribbling in public, I (and others) might make line comments through the GitHub interface on your commits to point out "things you might want to know".  You don't have to wait until you're all finished...!  There may be things that if known up front would change your approach.

**Comment by hostilefork** [comment]
I know you likely haven't been looking at this for a while _(and I've been taking a couple months off from Ren-C or Rebol-related things myself...)_  But I thought I would update this as I have a clearer picture now.

I've firmed up my beliefs that:

You cannot put "unsets" in blocks.  Further, to make things clearer, there is no such "datatype" as UNSET!, e.g. you can never say **type-of x** and get an UNSET! back.  Instead, **type-of ()** is a _none value_.  So you could write:

```
either type: type-of do code [
    print ["The return type of the evaluation was" type]
][
    print ["The evaluation returned no value."]
]
```

As there is no datatype representing unsets, the UNSET? type-checker is moved to <r3-legacy> and replaced with a new VOID? checker.  (But again, there is no VOID! type)

After considerable reflection on the issue, I've determined that prohibiting the assignment of voids to a SET-WORD! is not a valuable protection, compared to the usefulness of being able to clear out a variable by assigning it.  Rather, Ren-C has made it valuable to be able to work with expressions that are not set...since the new bias is that **if false [...]** returns a void instead of a NONE! value.  So this means if `get-env` returns a void when it doesn't find a value in the environment, you would be able to say:

```
val: get-env "QUERY_STRING"
if not void? :val [probe val]
```

In trying to reclaim the words "set?" and "unset?", a first step is that you could write the above as **if set? 'val**.  In the future, after some time has passed, UNSET? would make a comeback as something that you pass a word or a path, and it would be equivalent to **not set?**.

Similarly to the above, since UNSET!/VOID! is not a type, it is not in the ANY-TYPE! collection.  So instead of **not void? :val** you could say **any-type? :val**.  Moreover, in Ren-C the more sensible name **ANY-VALUE!** has been given to this typeclass, as "ANY-TYPE!" is too hard to distinguish from "ANY-DATATYPE!" (which sounds like it might mean WORD!, STRING!, etc. as a typeset covering any type...as opposed to any Rebol value...now meaning not a void).

In the further imaginations of the future, **value?** in its current form (which is something like **set?**) would be deprecated... and someday make a return as a convenient shorthand for **any-value?**.  So `if not void? :val` could be written as `if value? :val` when that came to pass.

But I still think that in Rebol's model, demonstrating at the time of fetch that you know it might be not there is the best idea.  Otherwise you write code like:

```
append data get-env "QUERY_STRING"
```

If it returns void, then in Ren-C's world it would just add nothing.  (Because **`append copy [a b c] ()`** is **`[a b c]`**.  But previously it would have appended a none.  I think the choice of adding nothing is better, but only if you show some awareness of it.

**SO IN CONCLUSION**: The right answer to this is probably:
- Make **get-env** error by default if the key is not there.
- Have **get-env/opt** return void, which is easy enough to work with 

What all of this spells is just part of the much more comfortable formulation of how "unsets" can be handled.  It means that one can be a bit less concerned about working with them, because they are not so "hot to handle"... the following isn't so bad, is it?

```
if value? val: get-env/opt "QUERY_STRING" [
    append data val
]
```

(Noting again that today you'd have to say that as **if not void?**)  And even better:

```
append data get-env/opt "QUERY_STRING"
```

Hopefully all of that makes sense.  :-)  This stuff is not committed, so I won't close it yet...but just wanted to make a note...

**Comment by hostilefork** [comment]
To whom it may concern :-)

The lore of "voids" and "blanks" has been evolving considerably.  Voids usefulness as an "out of band" non ANY-VALUE!...which you cannot find purposefully "stored" into a block or a variable...is powerful for those who care about the difference between the absence of a thing vs. a thing that's supposed to denote an absence.

The shift in Ren-C is now to export fundamental natives which have possibly funny names and make the distinction (e.g. `SELECT*`) and then create usermode wrappers which may have more convenient behavior (e.g. a SELECT function that turns the voids returned by the native SELECT* into blanks)

This distinction is now moving forward even to GET, so there is no GET/ANY or GET/OPT... instead the core is GET* which has the semantics of GET/OPT.  Then GET is the "convenience" function which blankifies void results, so **if blank? get 'x [print "that could mean x is either not set or it is set to blank"]**.  A stronger GET-VALUE can be used to have the semantics formerly of R3-Alpha's GET, which will error in the case of the variable being unset.

All of this points to the idea that something with a name like GET-ENV (if that stays its name, and is how these variables are retreived) should return blank.  And if `environment` became a kind of simulated object so you could write **environment/HOME**, then it would follow the rules of other objects.... if HOME is not defined then `environment/HOME` is an error (GET-VALUE semantics), while `:environment/HOME` is void (GET* semantics), while `get 'environment/home` is blank (GET semantics)

Long story short: GET-ENV will be blank for missing variables.  :-)
